coriolis/licensing/client.py

```diff
@@ -0,0 +1,188 @@
+
+    @utils.retry_on_error()
+    def _do_req(
+            self, method_name, resource, body=None,
+            response_key=None, raw_response=False):
+        method = getattr(requests, method_name.lower(), None)
+        if not method:
+            raise ValueError("No such HTTP method '%s'" % method_name)
+
+        url = self._get_url_for_resource(resource)
+
+        kwargs = {"verify": self._verify}
+        if body:
+            if not isinstance(body, (str, bytes)):
+                body = json.dumps(body)
+            kwargs["data"] = body
+
+        LOG.debug(
+            "Making '%s' call to licensing server at '%s' with body: %s",
+            method_name, url, kwargs.get('data'))
+        resp = method(url, **kwargs)
+
+        if raw_response:
+            return resp
+
+        if not resp.ok:
+            # try to extract error from licensing server:
+            error = None
+            try:
+                error = resp.json().get('error', {})
+            except (Exception, KeyboardInterrupt):
+                LOG.debug(
+                    "Exception occured during error extraction from licensing "
+                    "response: '%s'\nException:\n%s",
+                    resp.text, utils.get_exception_details())
+            if error and all([x in error for x in ['code', 'message']]):
+                raise exception.Conflict(
+                    message=error['message'],
+                    code=int(error['code']))
+            else:
+                resp.raise_for_status()
+
+        resp_data = resp.json()
+        if response_key:
+            if response_key not in resp_data:
+                raise ValueError(
+                    "No response key '%s' in response body: %s" % (
+                        response_key, resp_data))
+            resp_data = resp_data[response_key]
+
+        return resp_data
+
+    def _get(self, resource, response_key=None):
+        return self._do_req("GET", resource, response_key=response_key)
+
+    def _post(self, resource, body, response_key=None):
+        return self._do_req(
+            "POST", resource, body=body, response_key=response_key)
+
+    def _put(self, resource, body, response_key=None):
+        return self._do_req(
+            "PUT", resource, body=body, response_key=response_key)
+
+    def _delete(self, resource, body, response_key=None):
+        return self._do_req(
+            "DELETE", resource, body=body, response_key=response_key)
+
+    def get_licence_status(self):
+        """ Gets licence status for appliance. """
+        return self._get("/licence-status", "licence_status")
+
+    def get_licences(self):
+        """ Lists all installed licences. """
+        return self._get("/licences", response_key="licences")
+
+    def add_licence(self, licence_data):
+        """ Sends request to add licence (in .PEM format). """
+        return self._post("/licences", licence_data)
+
+    def add_reservation(self, typ, num_vms):
```

Comment:
typ --> type

No use in abbreviating for the sake of one letter.

Reply:
It's named that way so as to avoid shadowing the `type` builtin.

Reply:
then call it ```reservation_type```. Something a bit more meaningful.